PathPlanning/DynamicMovementPrimitives/dynamic_movement_primitives.py

```diff
@@ -0,0 +1,260 @@
+
+            dimension_data = training_data[:, dim]
+
+            q0 = init_state[dim]
+            g = goal_state[dim]
+
+            q = q0
+            qd_last = 0
+
+            phi_vals = []
+            f_vals = []
+
+            for i, _ in enumerate(dimension_data):
+                if i + 1 == len(dimension_data):
+                    qd = 0
+                else:
+                    qd = (dimension_data[i+1] - dimension_data[i]) / dt
+
+                Phi = [np.exp(-0.5 * ((i * dt / data_period) - c)**2 / H)
+                       for c in C]
+                Phi = Phi/np.sum(Phi)
+
+                qdd = (qd - qd_last)/dt
+
+                f = (qdd * data_period**2 - self.K * (g - q) + self.B * qd
+                     * data_period) / (g - q0)
+
+                phi_vals.append(Phi)
+                f_vals.append(f)
+
+                qd_last = qd
+                q += qd * dt
+
+            phi_vals = np.asarray(phi_vals)
+            f_vals = np.asarray(f_vals)
+
+            w = np.linalg.lstsq(phi_vals, f_vals, rcond=None)
+
+            if self.weights is None:
+                self.weights = np.asarray(w[0])
+            else:
+                self.weights = np.vstack([self.weights, w[0]])
+
+    def recreate_trajectory(self, init_state, goal_state, T):
+        """
+        init_state - initial state/position
+        goal_state - goal state/position
+        T  - amount of time to travek q0 -> g
+        """
+
+        nrBasis = len(self.weights[0])  # number of gaussian basis functions
+
+        # means (C) and std devs (H) of gaussian basis functions
+        C = np.linspace(0, 1, nrBasis)
+        H = (0.65*(1./(nrBasis-1))**2)
+
+        # initialize virtual system
+        t = 0
+
+        # for plotting
+        self.train_t_vals = np.arange(0, T, self.timesteps)
+
+        if not isinstance(init_state, np.ndarray):
+            init_state = np.asarray(init_state)
+        if not isinstance(goal_state, np.ndarray):
+            goal_state = np.asarray(goal_state)
+
+        q = init_state
+        dimensions = self.weights.shape[0]
+        qd = np.zeros(dimensions)
+
+        positions = []
+        for k in range(self.timesteps):
+            new_state = []
+            t = t + self.dt
+
+            qdd = np.zeros(dimensions)
+
+            for dim in range(dimensions):
+
+                if t <= T:
+                    Phi = [np.exp(-0.5 * ((t / T) - c)**2 / H) for c in C]
+                    Phi = Phi / np.sum(Phi)
+                    f = np.dot(Phi, self.weights[dim])
+                else:
+                    f = 0
+
+                # simulate dynamics
+                qdd[dim] = (self.K*(goal_state[dim] - q[dim])/T**2
+                            - self.B*qd[dim]/T
+                            + (goal_state[dim] - init_state[dim])*f/T**2)
+
+            qd = qd + qdd * self.dt
+            q = q + qd * self.dt
+
+            if not isinstance(q, list):
```

Comment:
`q` is not always `np.array`? If so I think the instance check is redundant.

Reply:
It would be on the first loop, I updated this function so everything is numpy arrays